### PR TITLE
ESDT transfer role, no global burn/mint

### PR DESCRIPTION
**Pull request by sasurobert**
No more global mint burn
ESDT transfer role

### CODE REVIEW

**Review by iulianpascalau — COMMENTED** [review]
please add in elrond-go/genesis/process/shardGenesisCreator.go L61:
```GO
BuiltInFunctionOnMetaEnableEpoch: unreachableEpoch,
```

**Review by raduchis — CHANGES_REQUESTED** [review]
small changes and clarifications needed

File: genesis/process/shardGenesisBlockCreator.go [hunk 1]
```diff
@@ -277,6 +277,9 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
 		ShardCoordinator:             arg.ShardCoordinator,
 		EpochNotifier:                epochNotifier,
 		ESDTMultiTransferEnableEpoch: unreachableEpoch,
+		ESDTTransferRoleEnableEpoch:  unreachableEpoch,
+		GlobalBurnMintDisableEpoch:   unreachableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:281]: this is genesis block processor.

File: factory/apiResolverFactory.go [hunk 2]
```diff
@@ -346,6 +355,9 @@ func createBuiltinFuncs(
 		ShardCoordinator:             shardCoordinator,
 		EpochNotifier:                epochNotifier,
 		ESDTMultiTransferEnableEpoch: esdtMultiTransferEnableEpoch,
+		ESDTTransferRoleEnableEpoch:  esdtTransferRoleEnableEpoch,
+		GlobalBurnMintDisableEpoch:   esdtGlobalMintBurnDisableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply factory/apiResolverFactory.go:359]: done

File: process/smartContract/builtInFunctions/factory.go [hunk 3]
```diff
@@ -58,6 +61,9 @@ func CreateBuiltInFunctionContainer(args ArgsCreateBuiltInFunctionContainer) (vm
 		ShardCoordinator:                    args.ShardCoordinator,
 		EpochNotifier:                       args.EpochNotifier,
 		ESDTNFTImprovementV1ActivationEpoch: args.ESDTMultiTransferEnableEpoch,
+		ESDTTransferToMetaEnableEpoch:       args.ESDTTransferMetaEnableEpoch,
+		ESDTTransferRoleEnableEpoch:         args.ESDTTransferRoleEnableEpoch,
+		GlobalBurnMintDisableEpoch:          args.GlobalBurnMintDisableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/factory.go:66]: done

File: process/transaction/metaProcess.go [hunk 4]
```diff
@@ -76,9 +79,10 @@ func NewMetaTxProcessor(args ArgsNewMetaTxProcessor) (*metaTxProcessor, error) {
 	baseTxProcess.flagPenalizedTooMuchGas.Unset()
 
 	txProc := &metaTxProcessor{
-		baseTxProcessor: baseTxProcess,
-		txTypeHandler:   args.TxTypeHandler,
-		esdtEnableEpoch: args.ESDTEnableEpoch,
+		baseTxProcessor:            baseTxProcess,
+		txTypeHandler:              args.TxTypeHandler,
+		esdtEnableEpoch:            args.ESDTEnableEpoch,
+		builtInFunctionEnableEpoch: args.BuiltInFunctionOnMetachainEnableEpoch,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/transaction/metaProcess.go:85]: done

File: process/smartContract/process.go [hunk 5]
```diff
@@ -182,32 +185,33 @@ func NewSmartContractProcessor(args ArgsNewSmartContractProcessor) (*scProcessor
...
+		builtinEnableEpoch:                    args.BuiltinEnableEpoch,
+		repairCallBackEnableEpoch:             args.RepairCallbackEnableEpoch,
+		penalizedTooMuchGasEnableEpoch:        args.PenalizedTooMuchGasEnableEpoch,
+		isGenesisProcessing:                   args.IsGenesisProcessing,
+		stakingV2EnableEpoch:                  args.StakingV2EnableEpoch,
+		returnDataToLastTransferEnableEpoch:   args.ReturnDataToLastTransferEnableEpoch,
+		senderInOutTransferEnableEpoch:        args.SenderInOutTransferEnableEpoch,
+		builtInFunctionOnMetachainEnableEpoch: args.BuiltInFunctionOnMetachainEnableEpoch,
+		arwenChangeLocker:                     args.ArwenChangeLocker,
+		vmOutputCacher:                        args.VMOutputCacher,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go:214]: done

File: vm/systemSmartContracts/esdt.go [hunk 6]
```diff
@@ -103,13 +108,15 @@ func NewESDTSmartContract(args ArgsNewESDTSmartContract) (*esdt, error) {
...
-		endOfEpochSCAddress:    args.EndOfEpochSCAddress,
-		addressPubKeyConverter: args.AddressPubKeyConverter,
+		ownerAddress:               []byte(args.ESDTSCConfig.OwnerAddress),
+		eSDTSCAddress:              args.ESDTSCAddress,
+		hasher:                     args.Hasher,
+		marshalizer:                args.Marshalizer,
+		enabledEpoch:               args.EpochConfig.EnableEpochs.ESDTEnableEpoch,
+		globalMintBurnDisableEpoch: args.EpochConfig.EnableEpochs.GlobalMintBurnDisableEpoch,
+		transferRoleEnabpeEpoch:    args.EpochConfig.EnableEpochs.ESDTTransferRoleEnableEpoch,
+		endOfEpochSCAddress:        args.EndOfEpochSCAddress,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:118]: done

File: vm/systemSmartContracts/esdt.go [hunk 7]
```diff
@@ -501,6 +515,11 @@ func getStringFromBool(val bool) string {
 }
 
 func (e *esdt) burn(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !e.flagGlobalMintBurn.IsSet() {
+		e.eei.AddReturnMessage("global burn is not more enabled, use local burn")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:519]: done

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -549,6 +568,11 @@ func (e *esdt) burn(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 }
 
 func (e *esdt) mint(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !e.flagGlobalMintBurn.IsSet() {
+		e.eei.AddReturnMessage("global mint is not more enabled, use local mint")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:572]: done

File: config/epochConfig.go [hunk 9]
```diff
@@ -46,6 +46,9 @@ type EnableEpochs struct {
 	WaitingListFixEnableEpoch                   uint32
 	IncrementSCRNonceInMultiTransferEnableEpoch uint32
 	ESDTMultiTransferEnableEpoch                uint32
+	GlobalMintBurnDisableEpoch                  uint32
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply config/epochConfig.go:49]: done

File: vm/systemSmartContracts/esdt.go [hunk 10]
```diff
@@ -251,7 +258,10 @@ func (e *esdt) issue(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 	}
 
 	initialSupply := big.NewInt(0).SetBytes(args.Arguments[2])
-	if initialSupply.Cmp(big.NewInt(0)) <= 0 {
+	if e.flagGlobalMintBurn.IsSet() && initialSupply.Cmp(zero) <= 0 {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:261]: done

**Review by iulianpascalau — APPROVED** [review]

File: factory/apiResolverFactory.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/apiResolverFactory.go]: ok

File: genesis/process/shardGenesisBlockCreator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply genesis/process/shardGenesisBlockCreator.go]: ok

File: process/smartContract/builtInFunctions/factory.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply process/smartContract/builtInFunctions/factory.go]: ok

File: vm/systemSmartContracts/esdt.go [hunk 11]
```diff
@@ -51,10 +51,15 @@ type esdt struct {
...
-	enabledEpoch           uint32
-	flagEnabled            atomic.Flag
 	mutExecution           sync.RWMutex
 	addressPubKeyConverter core.PubkeyConverter
+
+	enabledEpoch               uint32
+	flagEnabled                atomic.Flag
+	globalMintBurnDisableEpoch uint32
+	flagGlobalMintBurn         atomic.Flag
+	transferRoleEnabpeEpoch    uint32
```
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/esdt.go]: ok

**Review by raduchis — DISMISSED** [review]

**Review by raduchis — APPROVED** [review]

**Review by iulianpascalau — DISMISSED** [review]

**Review by gabi-vuls — APPROVED** [review]
image
System test passed.

File: integrationTests/testProcessorNode.go [hunk 12]
```diff
@@ -465,22 +461,16 @@ func NewTestProcessorNodeWithBLSSigVerifier(
...
+// NewTestProcessorNodeWithEnableEpochs returns a TestProcessorNode instance custom enable epochs
+func NewTestProcessorNodeWithEnableEpochs(
 	maxShards uint32,
 	nodeShardId uint32,
 	txSignPrivKeyShardId uint32,
-	builtinEnableEpoch uint32,
-	deployEnableEpoch uint32,
-	relayedTxEnableEpoch uint32,
-	penalizedTooMuchGasEnableEpoch uint32,
+	enableEpochs config.EnableEpochs,
```
Comment (iulianpascalau) [inline integrationTests/testProcessorNode.go:469]: :+1: :1st_place_medal:

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Pull request merged**